src/plugins/query_enhancements/public/datasets/s3_handler.ts

```diff
@@ -3,111 +3,247 @@
         };
-      case 'CONNECTION':
+      }
+      case 'CONNECTION': {
+        const databases = await fetchDatabases(http, path);
         return {
           ...dataStructure,
           columnHeader: 'Databases',
           hasNext: true,
-          children: [
-            {
-              id: `${dataStructure.id}.defaultDb`,
-              title: 'defaultDb',
-              type: 'DATABASE',
-            },
-          ],
+          children: databases,
         };
-      case 'DATABASE':
+      }
+      case 'DATABASE': {
+        const tables = await fetchTables(http, path);
         return {
           ...dataStructure,
           columnHeader: 'Tables',
           hasNext: false,
-          children: [
-            {
-              id: `${dataStructure.id}.table1`,
-              title: 'table1',
-              type: 'TABLE',
-            },
-            {
-              id: `${dataStructure.id}.table2`,
-              title: 'table2',
-              type: 'TABLE',
-            },
-          ],
+          children: tables,
         };
-      default:
-        const s3DataSources = await fetchS3DataSources(savedObjects);
+      }
+      default: {
+        const dataSources = await fetchDataSources(client);
         return {
           ...dataStructure,
-          columnHeader: 'S3 Data Sources',
-          hasNext: false,
-          children: s3DataSources,
+          columnHeader: 'Clusters',
+          hasNext: true,
+          children: dataSources,
         };
+      }
     }
   },
 
   fetchFields: async (dataset: Dataset): Promise<DatasetField[]> => {
-    // This is a placeholder. You'll need to implement the actual logic to fetch S3 fields.
-    // For now, we'll return an empty array.
     return [];
   },
 
-  supportedLanguages: (): string[] => {
-    return ['sql']; // Assuming S3 only supports SQL queries
+  supportedLanguages: (dataset: Dataset): string[] => {
+    return ['SQL'];
   },
 };
 
-const fetchS3DataSources = async (client: SavedObjectsClientContract): Promise<DataStructure[]> => {
-  return [];
+const fetch = (
+  http: HttpSetup,
+  path: DataStructure[],
+  type: 'DATABASE' | 'TABLE'
+): Promise<DataStructure[]> => {
+  return new Promise((resolve, reject) => {
+    const dataSource = path.find((ds) => ds.type === 'DATA_SOURCE');
+    const parent = path[path.length - 1];
+    const meta = parent.meta as DataStructureCustomMeta;
+
+    timer(0, 5000)
+      .pipe(
+        mergeMap(() =>
+          http.fetch('../../api/enhancements/datasource/jobs', {
+            query: {
+              id: dataSource?.id,
+              queryId: meta.query.id,
+            },
+          })
+        ),
+        takeWhile(
+          (response) => response.status !== 'SUCCESS' && response.status !== 'FAILED',
+          true
+        ),
+        filter((response) => response.status === 'SUCCESS'),
+        map((response) => {
+          if (response.status === 'FAILED') {
+            throw new Error('Job failed');
```

Comment:
failed response shouldn't get here after the filter?

Reply:
you are right will clean this up post merge.